datarobot_batch_scoring/network.py

```diff
@@ -218,66 +274,110 @@ def __iter__(self):
                     todo.append(batch2)
                     todo.sort()
 
-
-class Network(object):
-    def __init__(self, concurrency, timeout, ui=None):
-        self._executor = ThreadPoolExecutor(concurrency)
-        self._timeout = timeout
-        self._ui = ui or logger
-        self.futures = []
-        self.concurrency = concurrency
-
-        self.session = requests.Session()
-        adapter = requests.adapters.HTTPAdapter(
-            pool_connections=concurrency, pool_maxsize=concurrency)
-        self.session.mount('http://', adapter)
-        self.session.mount('https://', adapter)
-
-    def _request(self, request):
-        prepared = self.session.prepare_request(request)
-        try:
-            self.session.send(prepared, timeout=self._timeout)
-        except requests.exceptions.ReadTimeout:
-            self._ui.warning(textwrap.dedent("""The server did not send any data
-in the allotted amount of time.
-You might want to decrease the "--n_concurrent" parameters
-or
-increase "--timeout" parameter.
-"""))
-
-        except Exception as exc:
-            self._ui.debug('Exception {}: {}'.format(type(exc), exc))
-            try:
-                callback = request.kwargs['hooks']['response']
-            except AttributeError:
-                callback = request.hooks['response'][0]
-            response = FakeResponse(400, 'No Response')
-            callback(response)
-
-    def perform_requests(self, requests):
-        for r in requests:
-            while True:
-                self.futures = [i for i in self.futures if not i.done()]
-                if len(self.futures) < self.concurrency:
-                    self.futures.append(self._executor.submit(self._request,
-                                                              r))
-                    break
-                else:
-                    wait(self.futures, return_when=FIRST_COMPLETED)
-            yield
+    def request_cb(self, f):
+        futures = [i for i in self.futures if not i.done()]
+        self.ui.debug('cb {}: pending futures: {}'.format(f, len(futures)))
```

Comment:
cb - callback, it's native callback of future, that is called when future is actually done.

half-baked-not-so-async-ready handlers of requests library are run in target thread, before future is considered done. So we can't check number of pending futures in it.

So to get reliable detect of last one, we need to put it in real callback of ThreadPool's future.